### DISCUSSION TITLE
How to use external clojure libraries?

### DISCUSSION THREAD

**Discussion by omariosouto** [babashka/nbb#210] [Q&A] [answered]
Hey folks, does anyone tried to use some library like https://github.com/plumatic/schema ?

Is it actually possible?

**Comment by omariosouto** [comment]
I also put a question on that here: https://github.com/babashka/nbb/issues/20#issuecomment-1153079641

**Comment by omariosouto** [comment]
Actually I follow this: https://github.com/babashka/nbb/blob/main/doc/dependencies.md

And reach to this result:

<img width="1439" alt="image" src="https://user-images.githubusercontent.com/13791385/173220690-0dc53717-918b-4b01-8529-9246c9cb3297.png">

The repo is here: https://github.com/devsoutinho/app-labs/tree/main/_servers/webproxy

**Reply by borkdude** [reply]
Hey @omariosouto ! 

Recently schema became compatible with babashka which like nbb runs with the SCI interpreter. This required some changes in schema. Unfortunately the SCI interpreter cannot run all Clojure code constructs, but most of it. Some issues with the current version of schema + SCI is that SCI doesn't support `js*` (which I have a workaround for locally) but also that it relies on macros executed on the JVM (see `schema.utils/char-map` for example). Nbb runs fully in JavaScript. You will see the same issue if you are trying to run Schema with self-hosted ClojureScript (planck for example):

```
$ plk
ClojureScript 1.10.914
cljs.user=> (require '[schema.core])
Execution error (ExceptionInfo) at (<cljs repl>:1).
Invalid :refer, macro schema.utils/char-map does not exist in file /Users/borkdude/dev/schema/src/cljc/schema/utils.cljc
```

The easiest option out would be if nbb would offer schema as a built-in library which has a better chance of working since you don't have the SCI limitation (it would be pre-compiled using CLJS). But I'm thinking that maybe we should not add the entire CLJS ecosystem within the nbb npm library but perhaps split things into multiple npm libraries to keep downloads small (this would work better with AWS Lambda which has a 3MB limitation for console editing for example).

So we could support something like:

```
(require '[schema.core]) ;;=> demands that you have `@babashka/nbb-prismatic-schema` installed in your npm packages.
```

Nbb also supports building your own spin-off with your favorite libraries:

https://github.com/babashka/nbb/blob/main/doc/dev.md#features

Nbb is still pretty young so some of these things aren't fully fleshed out yet. Perhaps @logseq-cldwalker has some ideas here too.

**Comment by omariosouto** [comment]
@borkdude thanks for your answer! I will try to find an alternative, hope in future this work, will be amazing!!!

I'm actually doing some experiments creating web servers like this one: https://devsoutinho-webproxy.herokuapp.com/

The full code is here: https://github.com/devsoutinho/app-labs/tree/main/_servers/webproxy/webproxy/app

**Comment by borkdude** [comment, marked as answer]
@omariosouto I made something here. The prismatic/schema library is compiled to another npm dependency which you have to also include and then you can use schema. Please try it out and let me know how it works for you.

https://github.com/babashka/nbb#prismaticschema

**Comment by omariosouto** [comment]
Thanks for the update @borkdude!!! it works now, and I'm able to do the validations

I see that the `s/defn` is currently not working and `(s/set-fn-validation! false)` is not working as expected, but right now it allows me to do the code that I want, thanks!

Edit: I also try to run using `[schema.core :as s :include-macros true]`

<img width="821" alt="image" src="https://user-images.githubusercontent.com/13791385/173250139-b44e85d8-aae7-4db7-888a-43bf45891b06.png">

**Reply by borkdude** [reply]
Ah, yeah, we need to add special support for that, but it can be fixed. Please help me track these problems by creating individual issues for those.

**Reply by omariosouto** [reply]
I will create them right now \o

**Reply by omariosouto** [reply]
@borkdude  https://github.com/babashka/nbb/issues/215 https://github.com/babashka/nbb/issues/216 there they are, thanks for the awesome job!
